template/context.go

```diff
@@ -70,3 +78,175 @@ func (ctx *Context) Expand(s string) (string, error) {
 
 	return buf.String(), nil
 }
+
+// PayloadContext provides a context for expanding a config or secret payload.
+// NOTE: Be very careful adding any fields to this structure with types
+// that have methods defined on them. The template would be able to
+// invoke those methods.
+type PayloadContext struct {
+	Context
+	Secrets secretPayloadGetter
```

Comment:
I decided to avoid adding these to the base `Context` type because I wanted to avoid exposing secrets during template substitution on labels, environment variables, etc. This might be a desirable feature to have in its own right, but I was worried there might be more opportunities for injection here, so didn't want to bundle it into this PR.

@diogomonica: Any thoughts on where templates should be able to use secrets or not?

That said, I think when I change this to use the func map as you suggested, this structure will no longer be necessary. There will just be `Context` with the func map injected (in places where we allow secret/config substitution), and `Context` without the func map (where we disallow secret/config substitution). This is probably a cleaner result.